File: java/core/src/test/java/dev/deja/core/MatchTest.java
package dev.deja.core;

import dev.deja.core.cassette.JsonRpcMessage;
import dev.deja.core.match.JudgeVerdict;
import dev.deja.core.match.Match;
import dev.deja.core.match.SemanticJudge;
import dev.deja.core.match.SemanticMatchOptions;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;

class MatchTest {

    private JsonRpcMessage toolCall(Object id, String tool, Map<String, Object> extraParams) {
        Map<String, Object> params = new java.util.LinkedHashMap<>();
        params.put("name", tool);
        params.putAll(extraParams);
        return JsonRpcMessage.request(id, "tools/call", params);
    }

    // --- Matching Tier ---

    @Test
    void matchesIdenticalStructuresWhileIgnoringIds() {
        JsonRpcMessage incoming = toolCall(100, "test", Map.of());
        JsonRpcMessage recorded = toolCall(1, "test", Map.of());
        assertThat(Match.matchStructural(incoming, recorded)).isTrue();
    }

    @Test
    void ignoresMcpMetaParameters() {
        JsonRpcMessage incoming = toolCall(2, "test", Map.of("_meta", Map.of("progressToken", "123")));
        JsonRpcMessage recorded = toolCall(1, "test", Map.of());
        assertThat(Match.matchStructural(incoming, recorded)).isTrue();
    }

    @Test
    void normalizesEmptyParams() {
        JsonRpcMessage incoming = JsonRpcMessage.request(2, "resources/list", Map.of());
        JsonRpcMessage recorded = JsonRpcMessage.request(1, "resources/list", null);
        assertThat(Match.matchStructural(incoming, recorded)).isTrue();
    }

    @Test
    void failsStructuralMatchOnDifferentData() {
        JsonRpcMessage incoming = toolCall(2, "fetch", Map.of());
        JsonRpcMessage recorded = toolCall(1, "echo", Map.of());
        assertThat(Match.matchStructural(incoming, recorded)).isFalse();
    }

    @Test
    void isTolerantOfKeyOrder() {
        Map<String, Object> argsA = new java.util.LinkedHashMap<>();
        argsA.put("b", 2);
        argsA.put("a", 1);
        JsonRpcMessage incoming = toolCall(1, "fetch", Map.of("args", argsA));

        Map<String, Object> argsB = new java.util.LinkedHashMap<>();
        argsB.put("a", 1);
        argsB.put("b", 2);
        JsonRpcMessage recorded = toolCall(2, "fetch", Map.of("args", argsB));

        assertThat(Match.matchStructural(incoming, recorded)).isTrue();
    }

    // --- Hard safety gate ---

    @Test
    void neverMatchesToolsCallAcrossDifferentToolNamesEvenStructurallyClose() {
        JsonRpcMessage incoming = toolCall(1, "delete_file", Map.of("path", "/tmp/x"));
        JsonRpcMessage recorded = toolCall(2, "read_file", Map.of("path", "/tmp/x"));

        assertThat(Match.matchStructural(incoming, recorded)).isFalse();
        assertThat(Match.calculateSimilarity(incoming, recorded)).isZero();
    }

    @Test
    void neverMatchesAcrossDifferentMethods() {
        JsonRpcMessage incoming = JsonRpcMessage.request(1, "resources/list", null);
        JsonRpcMessage recorded = JsonRpcMessage.request(2, "tools/list", null);
        assertThat(Match.calculateSimilarity(incoming, recorded)).isZero();
    }

    @Test
    void neverFuzzyMatchesTwoDifferentPathsSharingALongLiteralPrefix() {
        // Regression test: found via real-server testing (TypeScript side) against the MCP
        // filesystem reference server. Two different files under the same long directory
        // prefix scored deceptively high on generic character/word similarity before this
        // gate existed, risking one file's content being confidently returned for a request
        // naming a different file.
        JsonRpcMessage incoming = toolCall(1, "read_text_file", Map.of("path", "/home/user/projects/deja/data/other.txt"));
        JsonRpcMessage recorded = toolCall(2, "read_text_file", Map.of("path", "/home/user/projects/deja/data/config.txt"));

        assertThat(Match.calculateSimilarity(incoming, recorded)).isZero();
    }

    @Test
    void gatesConservativelyOnAnySlashContainingStringEvenOrdinaryProse() {
        // The heuristic can't distinguish "this slash means a path" from "this slash means
        // prose", so it deliberately requires exact equality for any slash-containing value.
        JsonRpcMessage incoming = toolCall(1, "echo", Map.of("message", "yes/no answer please"));
        JsonRpcMessage recorded = toolCall(2, "echo", Map.of("message", "yes/no answer, please"));

        assertThat(Match.calculateSimilarity(incoming, recorded)).isZero();
    }

    @Test
    void stillFuzzyMatchesNearIdenticalProseWithNoPathLikeSeparators() {
        JsonRpcMessage incoming = toolCall(1, "echo", Map.of("message", "please say hello"));
        JsonRpcMessage recorded = toolCall(2, "echo", Map.of("message", "please say hellO"));

        assertThat(Match.calculateSimilarity(incoming, recorded)).isGreaterThan(0);
    }

    @Test
    void doesNotGateOnPathsThatAreIdentical() {
        JsonRpcMessage incoming = toolCall(1, "read_text_file", Map.of("path", "/a/b/c.txt", "extra", "x"));
        JsonRpcMessage recorded = toolCall(2, "read_text_file", Map.of("path", "/a/b/c.txt"));

        assertThat(Match.calculateSimilarity(incoming, recorded)).isGreaterThan(0);
    }

    // --- Semantic similarity primitives ---

    @Test
    void numericClosenessIsOneForEqualNumbersAndDecaysWithRelativeDistance() {
        assertThat(Match.numericCloseness(5, 5)).isEqualTo(1);
        assertThat(Match.numericCloseness(100, 101)).isGreaterThan(0.9);
        assertThat(Match.numericCloseness(1, 1000)).isLessThan(0.1);
    }

    @Test
    void trigramDiceIsOneForIdenticalStringsAndZeroForWhollyDissimilarOnes() {
        assertThat(Match.trigramDice("hello", "hello")).isEqualTo(1);
        assertThat(Match.trigramDice("hello world", "hxllo worlx")).isGreaterThan(0.5);
        assertThat(Match.trigramDice("abc", "xyz")).isZero();
    }

    @Test
    void tokenJaccardIgnoresTokenOrderAndRewardsOverlap() {
        assertThat(Match.tokenJaccard("fetch the url", "the url fetch")).isEqualTo(1);
        assertThat(Match.tokenJaccard("alpha beta", "gamma delta")).isZero();
    }

    // --- calculateSimilarity (semantic tier) ---

    @Test
    void returnsOneForAStructuralMatch() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "best pizza"));
        JsonRpcMessage recorded = toolCall(2, "search", Map.of("query", "best pizza"));
        assertThat(Match.calculateSimilarity(incoming, recorded)).isEqualTo(1);
    }

    @Test
    void scoresNearIdenticalParamsHighlyButNotIdentically() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "best pizza near me"));
        JsonRpcMessage recorded = toolCall(2, "search", Map.of("query", "best pizza near mee"));

        double score = Match.calculateSimilarity(incoming, recorded);
        assertThat(score).isGreaterThan(0.75).isLessThan(1);
    }

    @Test
    void scoresWildlyDifferentParamsLow() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "aaaaaaaaaa"));
        JsonRpcMessage recorded = toolCall(2, "search", Map.of("query", "zzzzzzzzzz completely unrelated"));
        assertThat(Match.calculateSimilarity(incoming, recorded)).isLessThan(0.5);
    }

    @Test
    void penalizesAMissingKeyRatherThanIgnoringIt() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "x", "extra", "y"));
        JsonRpcMessage recorded = toolCall(2, "search", Map.of("query", "x"));
        assertThat(Match.calculateSimilarity(incoming, recorded)).isLessThan(1);
    }

    // --- findSemanticMatch ---

    private record Candidate(String id, JsonRpcMessage msg) {
    }

    private <T> T await(CompletableFuture<T> future) {
        try {
            return future.get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    @Test
    void returnsTheHighestScoringCandidateAtOrAboveThreshold() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "best pizza near me"));
        List<Candidate> candidates = List.of(
                new Candidate("close", toolCall(2, "search", Map.of("query", "best pizza near mee"))),
                new Candidate("far", toolCall(3, "search", Map.of("query", "completely unrelated topic"))));

        SemanticMatchOptions<Candidate> options = SemanticMatchOptions.<Candidate>builder()
                .candidates(candidates)
                .messageExtractor(Candidate::msg)
                .threshold(0.75)
                .build();

        Optional<Candidate> match = await(Match.findSemanticMatch(incoming, options));
        assertThat(match).isPresent();
        assertThat(match.get().id()).isEqualTo("close");
    }

    @Test
    void returnsEmptyWhenNothingClearsTheThresholdAndThereIsNoJudge() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "alpha"));
        List<Candidate> candidates = List.of(
                new Candidate("far", toolCall(2, "search", Map.of("query", "completely different topic entirely"))));

        SemanticMatchOptions<Candidate> options = SemanticMatchOptions.<Candidate>builder()
                .candidates(candidates)
                .messageExtractor(Candidate::msg)
                .threshold(0.9)
                .build();

        assertThat(await(Match.findSemanticMatch(incoming, options))).isEmpty();
    }

    @Test
    void escalatesOnlyTheUncertainBandToTheJudgeNotEverythingBelowThreshold() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "best pizza near me"));
        Candidate uncertain = new Candidate("uncertain", toolCall(2, "search", Map.of("query", "best pizza near m")));
        Candidate hopeless = new Candidate("hopeless", toolCall(3, "search", Map.of("query", "completely unrelated topic")));

        List<JsonRpcMessage> judged = new java.util.ArrayList<>();
        SemanticJudge judge = (inc, recorded) -> {
            judged.add(recorded);
            return CompletableFuture.completedFuture(new JudgeVerdict(true));
        };

        SemanticMatchOptions<Candidate> options = SemanticMatchOptions.<Candidate>builder()
                .candidates(List.of(uncertain, hopeless))
                .messageExtractor(Candidate::msg)
                .threshold(0.95)
                .judge(judge)
                .build();

        Optional<Candidate> match = await(Match.findSemanticMatch(incoming, options));
        assertThat(match).isPresent();
        assertThat(match.get().id()).isEqualTo("uncertain");
        // The "hopeless" candidate's score is far below threshold - 0.15, so it must never
        // reach the judge.
        assertThat(judged).hasSize(1);
    }

    @Test
    void neverLetsAJudgeApproveAMatchAcrossDifferentToolNames() {
        JsonRpcMessage incoming = toolCall(1, "delete_file", Map.of("path", "/x"));
        Candidate wrongTool = new Candidate("wrong-tool", toolCall(2, "read_file", Map.of("path", "/x")));

        SemanticJudge judge = (inc, recorded) -> CompletableFuture.completedFuture(new JudgeVerdict(true));

        SemanticMatchOptions<Candidate> options = SemanticMatchOptions.<Candidate>builder()
                .candidates(List.of(wrongTool))
                .messageExtractor(Candidate::msg)
                .threshold(0.5)
                .judge(judge)
                .build();

        assertThat(await(Match.findSemanticMatch(incoming, options))).isEmpty();
    }

    @Test
    void failsClosedWhenTwoDifferentCandidatesTieForTheTopQualifyingScore() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "best pizza near me"));
        // Two distinct candidates with byte-identical message content score identically against
        // any given incoming message -- a guaranteed, reproducible tie, not a contrived mock.
        JsonRpcMessage tiedMsg = toolCall(2, "search", Map.of("query", "best pizza near mee"));
        List<Candidate> candidates = List.of(new Candidate("first", tiedMsg), new Candidate("second", tiedMsg));

        SemanticMatchOptions<Candidate> options = SemanticMatchOptions.<Candidate>builder()
                .candidates(candidates)
                .messageExtractor(Candidate::msg)
                .threshold(0.75)
                .build();

        assertThat(await(Match.findSemanticMatch(incoming, options))).isEmpty();
    }

    @Test
    void stillReturnsTheSoleWinnerWhenOnlyOneCandidateTiesItself() {
        JsonRpcMessage incoming = toolCall(1, "search", Map.of("query", "best pizza near me"));
        List<Candidate> candidates = List.of(
                new Candidate("close", toolCall(2, "search", Map.of("query", "best pizza near mee"))),
                new Candidate("far", toolCall(3, "search", Map.of("query", "completely unrelated topic"))));

        SemanticMatchOptions<Candidate> options = SemanticMatchOptions.<Candidate>builder()
                .candidates(candidates)
                .messageExtractor(Candidate::msg)
                .threshold(0.75)
                .build();

        Optional<Candidate> match = await(Match.findSemanticMatch(incoming, options));
        assertThat(match).isPresent();
        assertThat(match.get().id()).isEqualTo("close");
    }
}

File: ts/test/core/match.test.ts
import { describe, it, expect } from "vitest";
import { calculateSimilarity, findSemanticMatch, matchStructural, numericCloseness, tokenJaccard, trigramDice } from "../../src/core/match.js";
import { JsonRpcMessage } from "../../src/core/types.js";

describe("Matching Tier", () => {
  it("matches identical structures while ignoring IDs", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 100, method: "tools/call", params: { name: "test" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "test" } };

    expect(matchStructural(incoming, recorded)).toBe(true);
  });

  it("ignores MCP _meta parameters", () => {
    const incoming: JsonRpcMessage = {
      jsonrpc: "2.0",
      id: 2,
      method: "tools/call",
      params: { name: "test", _meta: { progressToken: "123" } },
    };
    const recorded: JsonRpcMessage = {
      jsonrpc: "2.0",
      id: 1,
      method: "tools/call",
      params: { name: "test" },
    };

    expect(matchStructural(incoming, recorded)).toBe(true);
  });

  it("normalizes empty params", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "resources/list", params: {} };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "resources/list" };

    expect(matchStructural(incoming, recorded)).toBe(true);
  });

  it("fails structural match on different data", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "fetch" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "echo" } };

    expect(matchStructural(incoming, recorded)).toBe(false);
  });

  it("is tolerant of key order (structural, not textual, equality)", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "fetch", args: { b: 2, a: 1 } } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { args: { a: 1, b: 2 }, name: "fetch" } };

    expect(matchStructural(incoming, recorded)).toBe(true);
  });
});

describe("hard safety gate", () => {
  it("never matches tools/call across different tool names, even structurally close", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "delete_file", path: "/tmp/x" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "read_file", path: "/tmp/x" } };

    expect(matchStructural(incoming, recorded)).toBe(false);
    expect(calculateSimilarity(incoming, recorded)).toBe(0);
  });

  it("never matches across different methods", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "resources/list" };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/list" };

    expect(calculateSimilarity(incoming, recorded)).toBe(0);
  });

  it("never fuzzy-matches two different paths sharing a long literal prefix", () => {
    // Regression test: found via real-server testing against server-filesystem. Two
    // different files under the same long directory prefix scored deceptively high on
    // generic character/word similarity before this gate existed, risking one file's
    // content being confidently returned for a request naming a different file.
    const incoming: JsonRpcMessage = {
      jsonrpc: "2.0",
      id: 1,
      method: "tools/call",
      params: { name: "read_text_file", path: "/home/user/projects/deja/data/other.txt" },
    };
    const recorded: JsonRpcMessage = {
      jsonrpc: "2.0",
      id: 2,
      method: "tools/call",
      params: { name: "read_text_file", path: "/home/user/projects/deja/data/config.txt" },
    };

    expect(calculateSimilarity(incoming, recorded)).toBe(0);
  });

  it("gates conservatively on any slash-containing string, even ordinary prose", () => {
    // The heuristic can't distinguish "this slash means a path" from "this slash means
    // prose", so it deliberately requires exact equality for any slash-containing value.
    // A missed fuzzy match (falls through to a clean "no match" error) is an acceptable
    // cost for guaranteeing a mismatched path/URI can never fuzzy-match.
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "echo", message: "yes/no answer please" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "echo", message: "yes/no answer, please" } };

    expect(calculateSimilarity(incoming, recorded)).toBe(0);
  });

  it("still fuzzy-matches near-identical prose that contains no path-like separators", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "echo", message: "please say hello" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "echo", message: "please say hellO" } };

    expect(calculateSimilarity(incoming, recorded)).toBeGreaterThan(0);
  });

  it("does not gate on paths that are identical", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "read_text_file", path: "/a/b/c.txt", extra: "x" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "read_text_file", path: "/a/b/c.txt" } };

    // Same path on both sides -- the gate must not fire just because a path-shaped value is
    // present; only a genuine mismatch between two path-shaped values should force 0.
    expect(calculateSimilarity(incoming, recorded)).toBeGreaterThan(0);
  });
});

describe("semantic similarity primitives", () => {
  it("numericCloseness is 1 for equal numbers and decays with relative distance", () => {
    expect(numericCloseness(5, 5)).toBe(1);
    expect(numericCloseness(100, 101)).toBeGreaterThan(0.9);
    expect(numericCloseness(1, 1000)).toBeLessThan(0.1);
  });

  it("trigramDice is 1 for identical strings and 0 for wholly dissimilar ones", () => {
    expect(trigramDice("hello", "hello")).toBe(1);
    expect(trigramDice("hello world", "hxllo worlx")).toBeGreaterThan(0.5);
    expect(trigramDice("abc", "xyz")).toBe(0);
  });

  it("tokenJaccard ignores token order and rewards overlap", () => {
    expect(tokenJaccard("fetch the url", "the url fetch")).toBe(1);
    expect(tokenJaccard("alpha beta", "gamma delta")).toBe(0);
  });
});

describe("calculateSimilarity (semantic tier)", () => {
  it("returns 1.0 for a structural match", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "fetch", url: "https://a.test" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "fetch", url: "https://a.test" } };

    expect(calculateSimilarity(incoming, recorded)).toBe(1);
  });

  it("scores near-identical params highly but not identically", () => {
    // Uses a prose-like "query" field rather than a path/URL: those are hard-gated to
    // exact-match below (two different paths are never "close enough" -- see the hard
    // safety gate tests), so this demonstrates genuine fuzzy scoring on a field where
    // near-equality is actually a meaningful signal.
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "search", query: "best pizza near me" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "search", query: "best pizza near mee" } };

    const score = calculateSimilarity(incoming, recorded);
    expect(score).toBeGreaterThan(0.75);
    expect(score).toBeLessThan(1);
  });

  it("scores wildly different params low", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "fetch", url: "https://example.com/aaa" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "fetch", url: "ftp://totally-different.internal/zzz" } };

    expect(calculateSimilarity(incoming, recorded)).toBeLessThan(0.5);
  });

  it("penalizes a missing key rather than ignoring it", () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "fetch", url: "https://a.test", extra: "x" } };
    const recorded: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "fetch", url: "https://a.test" } };

    expect(calculateSimilarity(incoming, recorded)).toBeLessThan(1);
  });
});

describe("findSemanticMatch", () => {
  interface Candidate {
    id: string;
    msg: JsonRpcMessage;
  }

  const getMessage = (c: Candidate) => c.msg;

  it("returns the highest-scoring candidate at or above threshold", async () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "search", query: "best pizza near me" } };
    const candidates: Candidate[] = [
      { id: "close", msg: { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "search", query: "best pizza near mee" } } },
      { id: "far", msg: { jsonrpc: "2.0", id: 3, method: "tools/call", params: { name: "search", query: "completely unrelated topic" } } },
    ];

    const match = await findSemanticMatch(incoming, { candidates, getMessage, threshold: 0.75 });
    expect(match?.id).toBe("close");
  });

  it("returns null when nothing clears the threshold and there is no judge", async () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "search", query: "alpha" } };
    const candidates: Candidate[] = [
      { id: "far", msg: { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "search", query: "completely different topic entirely" } } },
    ];

    const match = await findSemanticMatch(incoming, { candidates, getMessage, threshold: 0.9 });
    expect(match).toBeNull();
  });

  it("escalates only the uncertain band to the judge, not everything below threshold", async () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "search", query: "best pizza near me" } };
    const candidates: Candidate[] = [
      { id: "uncertain", msg: { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "search", query: "best pizza near m" } } },
      { id: "hopeless", msg: { jsonrpc: "2.0", id: 3, method: "tools/call", params: { name: "search", query: "completely unrelated topic" } } },
    ];

    const judged: JsonRpcMessage[] = [];
    const judge = async (_incoming: JsonRpcMessage, recorded: JsonRpcMessage) => {
      judged.push(recorded);
      return { equivalent: true };
    };

    const match = await findSemanticMatch(incoming, { candidates, getMessage, threshold: 0.95, judge });

    expect(match?.id).toBe("uncertain");
    // The "hopeless" candidate's score is far below threshold - 0.15, so it should never reach the judge.
    expect(judged).toHaveLength(1);
  });

  it("never lets a judge approve a match across different tool names (hard gate wins first)", async () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "delete_file", path: "/x" } };
    const candidates: Candidate[] = [
      { id: "wrong-tool", msg: { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "read_file", path: "/x" } } },
    ];

    const judge = async () => ({ equivalent: true }); // a judge that (wrongly) says yes to everything
    const match = await findSemanticMatch(incoming, { candidates, getMessage, threshold: 0.5, judge });

    expect(match).toBeNull();
  });

  it("fails closed (returns null) when two different candidates tie for the top qualifying score", async () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "search", query: "best pizza near me" } };
    // Two distinct candidates with byte-identical message content score identically against
    // any given incoming message -- a guaranteed, reproducible tie, not a contrived mock.
    const tiedMsg: JsonRpcMessage = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "search", query: "best pizza near mee" } };
    const candidates: Candidate[] = [
      { id: "first", msg: tiedMsg },
      { id: "second", msg: tiedMsg },
    ];

    const match = await findSemanticMatch(incoming, { candidates, getMessage, threshold: 0.75 });
    expect(match).toBeNull();
  });

  it("still returns the sole winner when only one candidate ties itself (no ambiguity)", async () => {
    const incoming: JsonRpcMessage = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "search", query: "best pizza near me" } };
    const candidates: Candidate[] = [
      { id: "close", msg: { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "search", query: "best pizza near mee" } } },
      { id: "far", msg: { jsonrpc: "2.0", id: 3, method: "tools/call", params: { name: "search", query: "completely unrelated topic" } } },
    ];

    const match = await findSemanticMatch(incoming, { candidates, getMessage, threshold: 0.75 });
    expect(match?.id).toBe("close");
  });
});
